torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -938,12 +1055,17 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
         grad, self.world_size * self._shard_size_multiple)
     if self.optimization_barrier_in_backward:
       self.optimization_barrier_op([grad_flat])
+    if grad_flat.dtype != torch.float32 and self.fp32_reduce_scatter:
+      grad_flat = grad_flat.to(torch.float32)
     reduced_grad = self.reduce_scatter_op(
         xm.REDUCE_SUM,
         grad_flat.detach(),
         scale=1.0,
         scatter_dim=0,
-        shard_count=self.world_size)
+        shard_count=self.world_size,
+        groups=self.sharding_groups)
+    if reduced_grad.dtype != torch.float32:
+      reduced_grad = reduced_grad.to(torch.float32)
```

Comment:
Are these two lines necessary?

If `fp32_reduce_scatter` is false, should we just keep a 16-bit data type and let the optimizer decide what to do? Or do existing prevailing optimizers don't like 16-bit grads?

Reply:
I think it would be better to cast the gradient back to fp32 here, because:
1) The model parameters are fp32, so one would naturally expect that the backward pass of a module would yield the same gradient dtype as the parameters. This is also how the MNIST and ImageNet examples can run with a simple change of `compute_dtype` flag -- the optimizer still sees fp32 gradients and can do the same optimization as before.
2) bfloat16 is very bad for gradient accumulation, e.g. 4096 + 1 = 4096 in bfloat16, and float16 also has less precision, so one would like to cast the gradients back to fp32 for accumulation (which is addressed in this cast `reduced_grad.to(torch.float32)`).
3) most existing optimizers like Adam needs to keep their 1st and 2nd momentum in fp32 (or at least the 2nd momentum in Adam) to work well and expect full precision gradients to accumulate the momentum.